### PR TITLE
Index iter bug

### PR DESCRIPTION
**Pull request by seberg**
Fixes the wrong indexing check due to bad iterator. I think this is a candidate for backporting.

I did not test if it is worth it to cache the value in the inner loop (not sure if it can make a difference anyway), but I doubt it.
Also added a warning to the documentation, it might be that grow inner loop flag has to be set as well to trigger this.

### CODE REVIEW

File: numpy/core/tests/test_indexing.py [hunk 1]
```diff
@@ -404,6 +404,22 @@ def __getitem__(self, item):
...
+        C = 4
+        x = np.ones(10)
+        x2 = np.ones((10, 2))
+        ind = np.arange(10)[:, None, None, None]
+        ind = np.broadcast_to(ind, (10, 55, C, C))
+
+        # single advanced index case
+        assert_array_equal(x[ind], x[ind.copy()])
+        # higher dimensional advanced index
+        zind = np.zeros(4, dtype=np.intp)
```
Comment (jaimefrio) [inline numpy/core/tests/test_indexing.py]: Wouldn't it make sense to use `C` instead of `4` here, like you do in the definition of `ind`?

Comment (seberg) [inline numpy/core/tests/test_indexing.py]: Just put in the 4, the C makes some sense when looking at the issue (and may really be an easy way to get the right array size), but I guess there is no big point here

Comment (charris) [inline numpy/core/tests/test_indexing.py]: `C`? We don't need no stinkin' `C`;

Comment (jaimefrio) [inline numpy/core/tests/test_indexing.py]: For the record, with `4`s everywhere it also LGTM.

**Pull request merged**